File: ros_nodes/mtf01_px4_bridge/test/test_flow.py
import math

import cv2
import numpy as np
import pytest

from mtf01_px4_bridge.flow import estimate_flow


def textured_image() -> np.ndarray:
    image = np.zeros((100, 100), dtype=np.uint8)
    for y in range(10, 91, 10):
        for x in range(10, 91, 10):
            cv2.circle(image, (x, y), 2, 255, -1)
    return image


def shifted(image: np.ndarray, x: float, y: float) -> np.ndarray:
    transform = np.float32([[1, 0, x], [0, 1, y]])
    return cv2.warpAffine(image, transform, (image.shape[1], image.shape[0]))


def test_estimates_px4_opencv_flow_convention():
    previous = textured_image()
    estimate = estimate_flow(previous, shifted(previous, 2.0, -3.0), math.radians(42))
    focal = 50.0 / math.tan(math.radians(21))

    assert estimate.quality > 0
    assert estimate.tracked_features >= 6
    assert estimate.integrated_x == pytest.approx(math.atan2(2.0, focal), abs=0.005)
    assert estimate.integrated_y == pytest.approx(math.atan2(-3.0, focal), abs=0.005)


def test_textureless_frame_has_zero_quality():
    image = np.zeros((100, 100), dtype=np.uint8)
    estimate = estimate_flow(image, image, math.radians(42))

    assert estimate.quality == 0
    assert estimate.integrated_x == 0.0
    assert estimate.integrated_y == 0.0

File: ros_nodes/mtf01_px4_bridge/test/test_flow.cpp
#include <cmath>

#include <gtest/gtest.h>
#include <opencv2/imgproc.hpp>

#include "mtf01_px4_bridge/flow.hpp"

namespace {

cv::Mat textured_image() {
  cv::Mat image = cv::Mat::zeros(100, 100, CV_8UC1);
  for (int y = 10; y <= 90; y += 10) {
    for (int x = 10; x <= 90; x += 10) {
      cv::circle(image, cv::Point(x, y), 2, cv::Scalar(255), cv::FILLED);
    }
  }
  return image;
}

cv::Mat shifted(const cv::Mat & image, double x, double y) {
  cv::Mat output;
  const cv::Mat transform = (cv::Mat_<double>(2, 3) << 1, 0, x, 0, 1, y);
  cv::warpAffine(image, output, transform, image.size());
  return output;
}

TEST(FlowEstimator, EstimatesPx4OpenCvFlowConvention) {
  const auto previous = textured_image();
  const auto estimate = mtf01_px4_bridge::estimate_flow(
      previous, shifted(previous, 2.0, -3.0), 42.0 * CV_PI / 180.0);
  const double focal = 50.0 / std::tan(21.0 * CV_PI / 180.0);

  EXPECT_GT(estimate.quality, 0);
  EXPECT_GE(estimate.tracked_features, 6);
  EXPECT_NEAR(estimate.integrated_x, std::atan2(2.0, focal), 0.005);
  EXPECT_NEAR(estimate.integrated_y, std::atan2(-3.0, focal), 0.005);
}

TEST(FlowEstimator, TexturelessFrameHasZeroQuality) {
  const cv::Mat image = cv::Mat::zeros(100, 100, CV_8UC1);
  const auto estimate = mtf01_px4_bridge::estimate_flow(image, image, 42.0 * CV_PI / 180.0);

  EXPECT_EQ(estimate.quality, 0);
  EXPECT_FLOAT_EQ(estimate.integrated_x, 0.0F);
  EXPECT_FLOAT_EQ(estimate.integrated_y, 0.0F);
}

}  // namespace
